src/VisualStudio/Core/Def/Implementation/Library/VsNavInfo/Extensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Library.VsNavInfo
+{
+    internal static class Extensions
+    {
+        public static void Add(this ImmutableArray<NavInfoNode>.Builder builder, string name, _LIB_LISTTYPE type, bool expandDottedNames)
+        {
+            if (name == null)
+            {
+                return;
+            }
+
+            if (expandDottedNames)
+            {
+                const char separator = '.';
+
+                var start = 0;
+                var separatorPos = name.IndexOf(separator, start);
+
+                while (separatorPos >= 0)
+                {
+                    builder.Add(name.Substring(start, separatorPos - start), type);
```

Comment:
Sure, I was trying to avoid allocating allocating two arrays unnecessarily -- the one for the params array and the one for the result.